docs!: Improve error message for unsupported VCS platform (bug 1793219)

catamorphism opened the pull request:
In parser.py, the UnsupportedPlatformError raised by parse() included `platform` rather than `parsed_url.platform`, leading to a misleading error message. This commit changes the error to use `parsed_url.host` instead, which is more helpful for diagnosing the error. It also changes `UnsupportedPlatformError` to take a list of supported platforms, which is helpful to the caller in printing out a helpful error message.

JohanLorenzo commented on mozilla_repo_urls/parser.py:
```diff
@@ -9,7 +9,7 @@
 
 
-_SUPPORTED_PLAFORMS = ("hgmo", "github")
+_SUPPORTED_PLATFORMS = ("hgmo", "github")
```
Thank you for fixing this nit!

JohanLorenzo commented on mozilla_repo_urls/errors.py:
```diff
@@ -8,8 +8,13 @@ def __init__(self, url_string) -> None:
         super().__init__(
-            f"Unsupported platform. Got: {platform}. "
-            f"Expected: {supported_platforms}. URL: {url_string}"
+            f"Unsupported version control host. Got: {platform}. "
```
Good call on communicating on hosts rather than platform. I agree it makes things easier to understand 👍 

Nit: per [1], we're now passing a host rather than a platform object. I'd suggest we rename `platform` into `host`, then. 

```suggestion
            f"Unsupported version control host. Got: {host}. "
```

[1] https://github.com/mozilla-releng/mozilla-repo-urls/pull/6/files#diff-3b1582b5ab46ad40a414ed592a20a14c85e6ccad8a4bef900dec491840691db1R31

JohanLorenzo commented on mozilla_repo_urls/errors.py:
```diff
@@ -8,8 +8,13 @@ def __init__(self, url_string) -> None:
-            f"Unsupported platform. Got: {platform}. "
-            f"Expected: {supported_platforms}. URL: {url_string}"
+            f"Unsupported version control host. Got: {platform}. "
+            f"Expected one of: {supported_platforms}. URL: {url_string}"
```
Nit: Similarly, we're now passing a list of supported hosts rather than platforms. Let's make that clear here.

```suggestion
            f"Expected one of: {supported_hosts}. URL: {url_string}"
```

JohanLorenzo commented on mozilla_repo_urls/errors.py:
```diff
@@ -8,8 +8,13 @@ def __init__(self, url_string) -> None:
 
 
 class UnsupportedPlatformError(RepoUrlsBaseError):
+    @property
+    def supported_platforms(self):
+        return list(self._supported_platforms)
```
Question: I'm sorry, I'm not sure to understand. What is the need for a property here? `supported_platforms`/`supported_hosts` is already a list and we could make it a public attribute, I think.

catamorphism replied:
Removed

JohanLorenzo commented on mozilla_repo_urls/errors.py:
```diff
@@ -8,8 +8,13 @@ def __init__(self, url_string) -> None:
+        return list(self._supported_platforms)
+
     def __init__(self, url_string, platform, supported_platforms) -> None:
+        self._supported_platforms = supported_platforms
```
From previous comment: if my understanding is correct, we can just make that attribute public.

```suggestion
        self.supported_hosts = supported_hosts
```

JohanLorenzo replied:
Circling back after reviewing https://github.com/taskcluster/taskgraph/pull/126#pullrequestreview-1140599675, we likely don't need this attribute anymore.

JohanLorenzo commented on mozilla_repo_urls/errors.py:
```diff
@@ -8,8 +8,13 @@ def __init__(self, url_string) -> None:
+    def supported_platforms(self):
+        return list(self._supported_platforms)
+
     def __init__(self, url_string, platform, supported_platforms) -> None:
```
From previous comments, let's update the function signature to get better namings: 

```suggestion
    def __init__(self, url_string, host, supported_hosts) -> None:
```

JohanLorenzo commented on mozilla_repo_urls/parser.py:
```diff
@@ -21,7 +21,22 @@ def parse(url_string):
+        raise UnsupportedPlatformError(
+            url_string,
+            parsed_url.host,
+            [
+                domain
+                for domains in [
+                    platform[1].DOMAINS
+                    for platform in giturlparse.platforms.PLATFORMS
+                    if platform[0] in _SUPPORTED_PLATFORMS
+                ]
+                for domain in domains
+            ],
```
Refactor: I love list comprehensions and I don't mind having nested ones. That said, I had some trouble to figure out what the final result looks like. I think we can extract that as a named module constant, since all values are already defined at the module level. What do you think? 

```suggestion
            _SUPPORTED_HOSTS,
```

JohanLorenzo commented on mozilla_repo_urls/parser.py:
```diff
@@ -9,7 +9,7 @@
 
-_SUPPORTED_PLAFORMS = ("hgmo", "github")
+_SUPPORTED_PLATFORMS = ("hgmo", "github")
 
```
Refactor: named module constant.

```suggestion
_SUPPORTED_HOSTS = tuple(sorted([
    host
    for domains in [
        platform[1].DOMAINS
        for platform in giturlparse.platforms.PLATFORMS
        if platform[0] in _SUPPORTED_PLATFORMS
    ]
    for host in domains
]))

```

We can make it a sorted tuple too, which should simplify testing. Speaking of testing, we can create a new file at `test/test_parser.py` with this content:

```
from mozilla_repo_urls import parser


def test_supported_hosts():
    assert parser._SUPPORTED_HOSTS == ("gist.github.com", "github.com", "hg.mozilla.org")
```

This way, it's simpler to see what hosts we actually support. How does that sound?

JohanLorenzo commented on mozilla_repo_urls/parser.py:
```diff
@@ -21,7 +21,22 @@ def parse(url_string):
-        raise UnsupportedPlatformError(url_string, platform, _SUPPORTED_PLAFORMS)
-
+    if parsed_url.platform not in _SUPPORTED_PLATFORMS:
+        """
+        For error reporting purposes, the exception object includes the domain
+        for each supported platform.
+        """
```
Question: I've never used triple double quotes for comments. What are the pros compared to a `# regular comment`? If I'm not mistaken, unlike a regular comment, this statement evaluates an unused string.

catamorphism replied:
It makes it easier to edit comments, since if line lengths change, then you don't need to adjust where each of the '#'s go, but it's not a big deal.

JohanLorenzo reviewed (commented):
Hi @catamorphism! 

Thank you very much for investigating then fixing the issue you've got with this brand new library. I appreciate the context you provided and the fix itself, which I believe makes things clearer!

By the way, I'm sorry for the review delay. I was on vacation for the past couple of weeks. 

I found a few nits worth fixing before merging this PR, in my opinion. Let me know if you would like to handle them. If not, that's totally fine - I understand it's already been 2 weeks since you opened this PR.

catamorphism commented:
Hi Johan - I think I've addressed all your comments. I don't know how to run the tests, though (it doesn't seem to be documented in the README).

Unfortunately, I'm going to be on medical leave from Monday until early January, so I won't be available to make other changes until then.

JohanLorenzo commented:
Thank you for making the changes @catamorphism! I'm sorry for the lack of documentation on that part. The repo is still fairly new. I wanted to make the tests run as part of CI but I haven't found the time yet.

I ran the tests locally (by running `tox`) and found a couple of nits. I just fixed them. I'll get a release out with your changes.

Thanks again for taking the time for this PR! I wish you well in your medical leave 🤞
